src/operator/tensor/indexing_op.h

```diff
@@ -80,84 +80,139 @@ inline bool EmbeddingOpType(const nnvm::NodeAttrs& attrs,
-    Shape2(oshape.ProdShape(0, oshape.ndim()-1), oshape[oshape.ndim()-1]), s);
-    Tensor<xpu, 2, DType> grad_in = outputs[1].get<xpu, 2, DType>(s);
-
-
-    if (req[embedding::kWeight] == kWriteTo || req[embedding::kWeight] == kAddTo) {
-      if (req[embedding::kWeight] == kWriteTo) {
-        grad_in = scalar<DType>(0.0f);
-      }
-      if ((grad_out.shape_[0] < grad_out.shape_[1]) && (grad_out.shape_[0] < 512)) {
-        AddTakeGrad(grad_in, data, grad_out);
-      } else {
-        Tensor<xpu, 2, int> workspace =
-          ctx.requested[embedding::kTempSpace].get_space_typed<xpu, 2, int>(
-            mshadow::Shape2(2, data.shape_.Size()), s);
-        Tensor<xpu, 1, int> sorted_data = workspace[0];
-        Tensor<xpu, 1, int> original_index = workspace[1];
-        sorted_data = tcast<int>(data);
-        original_index = range<int>(0, data.shape_.Size());
-        SortByKey(sorted_data, original_index, true);
-        AddTakeGradLargeBatch(grad_in, sorted_data, original_index, grad_out);
-      }
+inline bool TakeOpType(const nnvm::NodeAttrs& attrs,
+                       std::vector<int> *in_type,
+                       std::vector<int> *out_type) {
+  // using single dtype ("float32") for safety reason
+  CHECK_GE(in_type->size(), 2);
+  int dtype = (*in_type)[1];
+  CHECK_NE(dtype, -1) << "idx must have specified type";
+  for (index_t i = 0; i < in_type->size(); ++i) {
+    if ((*in_type)[i] == -1) {
+      (*in_type)[i] = dtype;
     } else {
-      LOG(FATAL) << "wrong req";
+      CHECK_EQ((*in_type)[i], dtype) << "This layer requires uniform type. "
+                                     << "Expected " << dtype << " v.s. given "
+                                     << (*in_type)[i];
     }
-  });
+  }
+  out_type->clear();
+  out_type->push_back(dtype);
+  return true;
+}
+
+template<typename xpu>
+void TakeOpForward(const nnvm::NodeAttrs& attrs,
+                   const OpContext& ctx,
+                   const std::vector<TBlob>& inputs,
+                   const std::vector<OpReqType>& req,
+                   const std::vector<TBlob>& outputs) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(req[indexing::kOut], kWriteTo);
+    CHECK_EQ(inputs.size(), 2);
+    CHECK_EQ(outputs.size(), 1);
+    CHECK_EQ(inputs[indexing::kData].ndim(), 2)
+        << "Indexing layer expects its array to be two-dimensional. "
+        << inputs[indexing::kData].ndim()
+        << " dimensional input is given instead";
+
+    const TShape& idxshape = inputs[indexing::kIdx].shape_;
+    const TShape& oshape = outputs[indexing::kOut].shape_;
+
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+        Tensor<xpu, 1, DType> idx = inputs[indexing::kIdx].get_with_shape<xpu, 1, DType>(
+            Shape1(idxshape.ProdShape(0, idxshape.ndim())), s);
+        Tensor<xpu, 2, DType> data = inputs[indexing::kData].get<xpu, 2, DType>(s);
```

Comment:
The current implementation should be able to support larger ndims.
For example, we can take along `axis=0` from an array that has the shape (4, 5, 6, 7).
To do this, we can change this line to
```
Tensor<xpu, 2, DType> data = inputs[indexing::kData].FlatTo2D<xpu, DType>(s);
```
This will automatically collapse the trailing dimensions.